src/rules/noSubmoduleImportsRule.ts

```diff
@@ -0,0 +1,115 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    isCallExpression,
+    isExternalModuleReference,
+    isIdentifier,
+    isImportDeclaration,
+    isImportEqualsDeclaration,
+    isTextualLiteral,
+} from "tsutils";
+import * as ts from "typescript";
+import * as Lint from "../index";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        ruleName: "no-submodule-imports",
+        description: Lint.Utils.dedent`
+            Disallows importing any submodule.`,
+        rationale: Lint.Utils.dedent`
+            Submodules of some packages are treated as private APIs and the import
+            paths may change without deprecation periods. It's best to stick with
+            top-level package exports.`,
+        optionsDescription: "A list of packages whose submodules are whitelisted.",
+        options: {
+            type: "array",
+            items: {
+                type: "string",
+            },
+            minLength: 0,
+        },
+        optionExamples: [true, [true, "rxjs", "@angular/core"]],
+        type: "functionality",
+        typescriptOnly: false,
+    };
+
+    public static FAILURE_STRING = "Submodule import paths from this package are disallowed; import from the root instead";
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithWalker(new NoSubmoduleImportsWalker(sourceFile, this.ruleName, this.ruleArguments));
+    }
+}
+
+class NoSubmoduleImportsWalker extends Lint.AbstractWalker<string[]> {
+    public walk(sourceFile: ts.SourceFile) {
+        const findDynamicImport = (node: ts.Node): void => {
+            if (isCallExpression(node) && node.arguments.length === 1 &&
+                (isIdentifier(node.expression) && node.expression.text === "require" ||
+                node.expression.kind === ts.SyntaxKind.ImportKeyword)) {
+                this.checkForBannedImport(node.arguments[0]);
+            }
+            return ts.forEachChild(node, findDynamicImport);
+        };
+
+        for (const statement of sourceFile.statements) {
+            if (isImportDeclaration(statement)) {
+                this.checkForBannedImport(statement.moduleSpecifier);
+            } else if (isImportEqualsDeclaration(statement)) {
+                if (isExternalModuleReference(statement.moduleReference) && statement.moduleReference.expression !== undefined) {
+                    this.checkForBannedImport(statement.moduleReference.expression);
+                }
+            } else {
+                ts.forEachChild(statement, findDynamicImport);
+            }
+        }
+    }
+
+    private checkForBannedImport(expression: ts.Expression) {
```

Comment:
this is a static function (doesn't use `this`), you can make it `static` or move to a plain helper function